### PR TITLE
fix(file-viewer): prevent scroll jump to top during Mothership streaming

### PR DESCRIPTION
**Pull request by waleedlatif1**
## Summary
- Fix root cause of scroll jump: `MarkdownCheckboxCtx.Provider` was conditionally rendered around the scroll container, causing it to unmount/remount (and reset `scrollTop` to 0) every time streaming started
- Add `useScrollAnchor` hook with a spacer element that inflates `scrollHeight` to prevent the browser from clamping `scrollTop` when streamed content temporarily shrinks
- Linger `activeSessionId` on `complete` so `streamingContent` never becomes `undefined` between two consecutive Mothership tool calls on the same file
- Gate session activation in `upsert` on the incoming session having renderable content, preventing a blank flash when a new session starts before its first chunk arrives
- Fix `shouldShowStreamingFilePanel` to keep the panel mounted while the lingered session is active
- Fix `use-chat` post-write navigation condition to work with lingered completed sessions
- Fix `useAutoScroll` to check proximity before pinning to bottom on stream start (prevents unwanted jump-to-bottom in the chat panel)

## Type of Change
- [x] Bug fix

## Testing
Tested manually — scrolled to middle of a file, asked Mothership to make multi-step edits. Scroll position preserved across tool call boundaries. Also verified single `patch`, fresh `write`, and cross-file edits all behave correctly.

Unit tests added:
- `use-scroll-anchor.test.ts`: 15 tests for `computeSpacerShortage` pure function
- `use-file-preview-sessions.test.tsx`: 7 new tests for linger, content-gate, and edge cases
- `text-editor-state.test.ts`: 5 new tests for inter-session content shrink scenario

## Checklist
- [x] Code follows project style guidelines
- [x] Self-reviewed my changes
- [x] Tests added/updated and passing
- [x] No new warnings introduced
- [x] I confirm that I have read and agree to the terms outlined in the [Contributor License Agreement (CLA)](./CONTRIBUTING.md#contributor-license-agreement-cla)

### CODE REVIEW

File: apps/sim/app/workspace/[workspaceId]/home/hooks/use-file-preview-sessions.ts [hunk 1]
```diff
@@ -106,15 +106,24 @@ export function reduceFilePreviewSessions(
...
+        const currentActive = state.activeSessionId ? nextSessions[state.activeSessionId] : null
+        const currentHasContent = currentActive
+          ? hasRenderableFilePreviewContent(currentActive)
+          : false
+        const incomingHasContent = hasRenderableFilePreviewContent(action.session)
+        nextActiveSessionId =
+          currentHasContent && !incomingHasContent ? state.activeSessionId : action.session.id
       }
+
+      return { sessions: nextSessions, activeSessionId: nextActiveSessionId }
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/app/workspace/[workspaceId]/home/hooks/use-file-preview-sessions.ts:125]: Fixed in f021927396. Both `hydrate` and `upsert(activate:false)` now use `pickActiveSessionId(...) ?? state.activeSessionId` — if `pickActiveSessionId` returns null (no non-complete successor) the lingered completed session is preserved instead of cleared to null. Same pattern as the `complete` case.

File: apps/sim/hooks/use-scroll-anchor.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    )
+
+    // Inflate spacer so scrollHeight >= needed, preventing scrollTop clamping.
+    if (spacer) spacer.style.minHeight = `${shortage}px`
+
+    // Restore scroll position (now valid because spacer ensures enough height).
+    if (el.scrollTop < targetScrollTop) {
+      el.scrollTop = targetScrollTop
+    }
+  }, [content])
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/hooks/use-scroll-anchor.ts]: Fixed in f021927396. Added `isStreaming` to the `useLayoutEffect` dependency array and changed the early-return guard to `!hasUserScrolledRef.current || !isStreaming` — so the spacer is cleared to 0 as soon as streaming ends, regardless of whether `content` changed. During streaming the inflation logic runs normally; after streaming ends the spacer is always cleared.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/sim/hooks/use-scroll-anchor.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+export function computeSpacerShortage(
+  targetScrollTop: number,
+  clientHeight: number,
+  scrollHeight: number,
+  prevSpacerHeight: number
+): number {
+  const needed = targetScrollTop + clientHeight
+  const naturalScrollHeight = scrollHeight - prevSpacerHeight
+  return Math.max(0, needed - naturalScrollHeight)
+}
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/hooks/use-scroll-anchor.ts:18]: Valid bug — fixed in 3b9923469d. The spacer inflates to exactly `targetScrollTop + clientHeight`, so after `el.scrollTop = targetScrollTop` the browser fires `onScroll` synchronously (during the reflow, as noted in the existing comment) with `distanceFromBottom = 0`. The re-engage condition was triggering on that artificial proximity, clearing `hasUserScrolledRef` and letting the next content update clear the spacer and jump the user to top.

Fix: added `spacerHeightRef` (mirrors the spacer's current minHeight without a DOM read). The re-engage check in `onScroll` now gates on `spacerHeightRef.current === 0` — re-engagement is blocked while the spacer is active. The spacer effect keeps `spacerHeightRef` in sync (set to `shortage` when inflating, `0` when clearing). When the user genuinely scrolls to the real document bottom with no active spacer, re-engagement fires correctly.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review
